File: app/src/lib/aa.ts
// Alternate Advancement helpers shared by the AA tab and the Spellbook tab.
// The mutation helpers write `build.aa_ranks` — the ONE source of truth for AA ranks.
import type { AaAbility, BuildInput } from "./api";

/** The AA that grants spell gems (8 base + 1 per rank, 6 ranks -> 14 gems). */
export const MNEMONIC_AA = "Mnemonic Retention";

/** Category sections, in display order. */
export const AA_CATEGORIES = ["GENERAL", "ARCHETYPE", "CLASS", "SPECIAL"] as const;

export function findAaByName(aas: AaAbility[], name: string): AaAbility | null {
  return aas.find((a) => a.name === name) ?? null;
}

/** Purchased rank of one AA (absent key = 0). */
export function aaRank(build: BuildInput, id: number): number {
  return build.aa_ranks?.[id] ?? 0;
}

/**
 * Set a rank, clamped to 0..max_rank. Rank 0 DELETES the key (a stored zero would
 * be noise in the saved build) and every write hands back a fresh object so the
 * resolve pipeline sees the change.
 */
export function setAaRank(build: BuildInput, aa: AaAbility, rank: number): void {
  const clamped = Math.min(aa.max_rank, Math.max(0, Math.floor(rank)));
  const next = { ...(build.aa_ranks ?? {}) };
  if (clamped === 0) delete next[aa.id];
  else next[aa.id] = clamped;
  build.aa_ranks = next;
  // The engine reads max(planner, legacy field), so a legacy value would fight a
  // lowered stepper. Once the planner owns Mnemonic Retention, retire the old field.
  if (aa.name === MNEMONIC_AA) build.aa_mnemonic_retention = 0;
}

/**
 * Cost of holding `rank` ranks: the wiki's per-rank costs are CUMULATIVE, so this is
 * the sum of the first N entries. Mirrors eql_engine::aa::rank_cost — a "?" rank is
 * unknowable, never free, so it is counted separately and NEVER folded in as 0.
 */
export function aaCost(aa: AaAbility, rank: number): { points: number; unknown: number } {
  let points = 0;
  let unknown = 0;
  for (let i = 0; i < Math.min(rank, aa.max_rank); i++) {
    const c = aa.costs[i];
    if (c == null) unknown++;
    else points += c;
  }
  return { points, unknown };
}

/** "12" when every rank is known, "4+?" / "?" when the wiki left ranks blank. */
export function fmtAaCost(aa: AaAbility, rank: number): string {
  const { points, unknown } = aaCost(aa, rank);
  if (unknown === 0) return String(points);
  return points === 0 ? "?" : `${points}+?`;
}

/** CLASS AAs are gated by the build's classes; the other categories are offered to all. */
export function aaGrantedByClasses(aa: AaAbility, classes: string[]): boolean {
  if (aa.category !== "CLASS") return true;
  const abbr = aa.class_abbr?.toUpperCase() ?? null;
  return abbr != null && classes.some((c) => c.toUpperCase() === abbr);
}

File: app/crates/eql-engine/src/aa.rs
//! Alternate Advancement planning (wiki "Alternate Advancement" page).
//!
//! Availability: General = every class; Class = only if you took that class;
//! Archetype = "depends on your class combo", but the wiki does NOT say which combo
//! grants which — so we show them all and flag that gap rather than invent a mapping.
//! Special = its own tab.
//!
//! Costs: the wiki's cost lists contain "?" for unfinished entries. A plan touching
//! any of those reports `cost_is_lower_bound`, so the UI never implies false precision.
use crate::Snapshot;
use eql_data::{AaPlan, BuildInput};

/// Is this AA offered to the build at all? (class gating only — level is reported,
/// not hidden, so a plan can look ahead.)
pub fn aa_available(aa: &eql_data::AaAbility, classes: &[String]) -> bool {
    match aa.category.as_str() {
        "CLASS" => aa
            .class_abbr
            .as_deref()
            .is_some_and(|c| classes.iter().any(|b| b.eq_ignore_ascii_case(c))),
        // GENERAL / SPECIAL: everyone. ARCHETYPE: the wiki doesn't publish the
        // class->archetype-AA mapping, so we can't filter it honestly.
        _ => true,
    }
}

/// Cost of taking `rank` ranks (ranks are cumulative: sum of the first N entries).
/// Returns (points, all_costs_known).
pub fn rank_cost(aa: &eql_data::AaAbility, rank: u32) -> (u32, bool) {
    let mut sum = 0;
    let mut known = true;
    for i in 0..rank.min(aa.max_rank) as usize {
        match aa.costs.get(i) {
            Some(Some(c)) => sum += c,
            _ => known = false, // wiki wrote "?" — unknowable, not free
        }
    }
    (sum, known)
}

pub fn resolve_aa(snapshot: &Snapshot, build: &BuildInput) -> AaPlan {
    let classes: Vec<String> = build.classes.iter().map(|c| c.to_uppercase()).collect();
    let mut plan = AaPlan {
        points_available: build.aa_points_available,
        ..Default::default()
    };
    for (id, rank) in &build.aa_ranks {
        let Some(aa) = snapshot.aas.iter().find(|a| a.id == *id) else { continue };
        let rank = (*rank).min(aa.max_rank);
        if rank == 0 {
            continue;
        }
        let (cost, known) = rank_cost(aa, rank);
        plan.points_spent += cost;
        if !known {
            plan.cost_is_lower_bound = true;
        }
        if aa.required_level.is_some_and(|lv| lv > build.level) {
            plan.level_locked.push(format!(
                "{} (needs level {})",
                aa.name,
                aa.required_level.unwrap_or(0)
            ));
        }
        if !aa_available(aa, &classes) {
            plan.class_locked.push(aa.name.clone());
        }
    }
    plan
}

/// Spell gems = 8 base + rank of the AA "Mnemonic Retention" (6 ranks max -> 14).
/// Reads the planner first, then the legacy standalone field (older saved builds).
pub fn spell_gems(snapshot: &Snapshot, build: &BuildInput) -> usize {
    let planned = snapshot
        .aas
        .iter()
        .find(|a| a.name == "Mnemonic Retention")
        .and_then(|a| build.aa_ranks.get(&a.id).copied())
        .unwrap_or(0);
    eql_data::spell_gem_count(planned.max(build.aa_mnemonic_retention))
}
